src/main/java/org/javarosa/core/model/instance/ExternalDataInstanceSource.java

```diff
@@ -1,64 +1,86 @@
     private String instanceId;
-    private Multimap<String, String> requestData;
     private boolean useCaseTemplate;
+    private String reference;
 
-    /**
-     * Externalizable constructor
-     */
-    public ExternalDataInstanceSource() {}
+    @Nullable
+    private String sourceUri;
+    private Multimap<String, String> requestData;
 
-    public ExternalDataInstanceSource(String instanceId, String sourceUri, Multimap<String, String> requestData, boolean useCaseTemplate) {
-        this(instanceId, null, sourceUri, requestData, useCaseTemplate);
-    }
+    @Nullable
+    private UUID storageReferenceId;
 
-    public ExternalDataInstanceSource(String instanceId, TreeElement root, String sourceUri, boolean useCaseTemplate) {
-        this(instanceId, root, sourceUri, ImmutableMultimap.of(), useCaseTemplate);
+    public final static String JR_REMOTE_REFERENCE = "jr://instance/remote";
+
+    public ExternalDataInstanceSource() {
     }
 
-    public ExternalDataInstanceSource(String instanceId, TreeElement root, String sourceUri, Multimap<String, String> requestData, boolean useCaseTemplate) {
+    private ExternalDataInstanceSource(String instanceId, TreeElement root, String reference, boolean useCaseTemplate,
+                                       String sourceUri, Multimap requestData, UUID storageReferenceId) {
         this.instanceId = instanceId;
-        this.sourceUri = sourceUri;
-        this.requestData = requestData;
         this.root = root;
+        this.reference = reference;
         this.useCaseTemplate = useCaseTemplate;
+        this.sourceUri = sourceUri;
+        this.requestData = requestData;
+        this.storageReferenceId = storageReferenceId;
     }
 
     /**
      * Copy constructor
      */
-    public ExternalDataInstanceSource(ExternalDataInstanceSource externalDataInstanceSource)  {
+    public ExternalDataInstanceSource(ExternalDataInstanceSource externalDataInstanceSource) {
         this.instanceId = externalDataInstanceSource.instanceId;
-        this.sourceUri = externalDataInstanceSource.sourceUri;
         this.root = externalDataInstanceSource.root;
+        this.reference = externalDataInstanceSource.reference;
         this.useCaseTemplate = externalDataInstanceSource.useCaseTemplate();
+        this.sourceUri = externalDataInstanceSource.sourceUri;
         this.requestData = externalDataInstanceSource.requestData;
+        this.storageReferenceId = externalDataInstanceSource.storageReferenceId;
+    }
+
+    public static ExternalDataInstanceSource buildRemoteDataInstanceSource(
+            String instanceId, @Nullable TreeElement root,
+            boolean useCaseTemplate, String sourceUri,
+            Multimap requestData) {
+        return new ExternalDataInstanceSource(instanceId,root, JR_REMOTE_REFERENCE,
+                useCaseTemplate, sourceUri, requestData, null);
+    }
+
+
+    public static ExternalDataInstanceSource buildStorageBackedDataInstanceSource(
```

Comment:
I think this could always take an `ExternalDataInstance` and a reference like `org.commcare.util.screen.MultiSelectEntityScreen#buildSelectedValuesInstance`

Reply:
That would mean that we would always need to create instance twice once without source and then with the source returned from here, one such place for ex  would be in [VirtualDataInstanceService](https://github.com/dimagi/formplayer/blob/refactorMultiSelect/src/main/java/org/commcare/formplayer/services/VirtualDataInstanceService.java#L63) where having this definition is more convenient.

Reply:
Ah true, I did look at that but of course it's not the same type. Fine to leave as is.